### PR TITLE
comments/experiments about Lab03

### PR DESCRIPTION
**Pull request by andreastedile**
As suggested by the instructor for the sake of exercise, I re-implemented the exercises of Lab 03 in PySMT.
I made the following unfortunate findings; I hope they will be useful for others:

- `Div` does not work (at least using the MathSAT solver). In facts, I cannot see it [there](https://github.com/pysmt/pysmt/blob/master/pysmt/solvers/msat.py).
- There is no way of doing the modulo operation to obtain the remainder of a division. The work of [this 2018 pull request](https://github.com/pysmt/pysmt/pull/476) does not seem to be moving forward.
- Sometimes (see `geometry.py`), PySMT raises the following exception: `pysmt.exceptions.NonLinearError`. This is because, at the moment, PySMT does not yet support MathSAT's Non-Linear Integer Arithmetic (NIA) capabilities; see PySMT's [Solvers Support](https://github.com/pysmt/pysmt#solvers-support). To overcome this annoyance, a friend of mine (@masinag) suggests to apply the following quick-fix:
Rewrite the [`walk_times`](https://github.com/pysmt/pysmt/blob/master/pysmt/solvers/msat.py#L975) function as follows:
```
    def walk_times(self, formula, args, **kwargs):
        res = args[0]
        nl_count = 0 if mathsat.msat_term_is_number(self.msat_env(), res) else 1
        for x in args[1:]:
            if not mathsat.msat_term_is_number(self.msat_env(), x):
                nl_count += 1
            # if nl_count >= 2:
            #     raise NonLinearError(formula)
            # else:
            #     res = mathsat.msat_make_times(self.msat_env(), res, x)
            res = mathsat.msat_make_times(self.msat_env(), res, x)
        return res
```

### CODE REVIEW

File: Lab03/students-proposal/220930/geometry.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+xa = Symbol("Xa", REAL)
+ya = Symbol("Ya", REAL)
+xb = Symbol("Xb", REAL)
+yb = Symbol("Yb", REAL)
+
+x = Symbol("x", REAL)
+m = Symbol("m", REAL)
+q = Symbol("q", REAL)
+
+f = Symbol("f", FunctionType(REAL, (REAL,)))
```
Comment (andreastedile) [inline Lab03/students-proposal/220930/geometry.py:36]: Declaration of uninterpreted function. Is this correct?

File: Lab03/students-proposal/220930/geometry.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    # This produces the following exception:
+    # pysmt.exceptions.NonLinearError: (m * (Xb - Xa))
+    # To bypass the exception, edit the "walk_times" function in pysmt/solvers as follows:
+    # def walk_times(self, formula, args, **kwargs):
+    #     res = args[0]
+    #     nl_count = 0 if mathsat.msat_term_is_number(self.msat_env(), res) else 1
+    #     for x in args[1:]:
+    #         res = mathsat.msat_make_times(self.msat_env(), res, x)
+    #     return res
+    msat.add_assertion(Equals(f(x), Plus(Times(m, x), q)))
```
Comment (andreastedile) [inline Lab03/students-proposal/220930/geometry.py:50]: This fixes the interpretation of the function `f` declared above.
Does PySMT allow to achieve the same with a one-liner?

File: Lab03/students-proposal/220930/geometry.py (hunk 1 above)
Comment (giuspek) [reply Lab03/students-proposal/220930/geometry.py:36]: Yes, this declaration works. FunctionType creates the type of your function, then the Function operator from pysmt.shortcuts must be used to call the function and set the arguments for each assertion.

**Comment by giuspek** [comment]
Hi Andrea,

all the considerations you reported are of interest to everyone, so I'll briefly mention everything in class. While the problem of non-linearity is partially justified (even though MathSAT will support non-linearity in the near future), it's really strange that the MathSAT class misses some important operations such as modulo and Div. It could be the case to fork the project and extend it (but only for personal purposes, this won't affect the exam and classes).

**Pull request merged**